src/gui/dbsettings/DatabaseSettingsWidgetStatistics.cpp

```diff
@@ -0,0 +1,207 @@
+            countGroup(*db.rootGroup());
+        }
+
+        // Get average password length
+        int pwdAvgLen() const
+        {
+            return pwds_.empty() ? 0 : pwdTotalLen / pwds_.size();
+        }
+
+        // Get max number of password reuse (=how many entries
+        // share the same password)
+        int maxPwdReuse() const
+        {
+            int ret = 0;
+            for (const auto& p : pwds_) {
+                ret = std::max(ret, p.second);
+            }
+            return ret;
+        }
+
+        // A warning sign is displayed if one of the
+        // following returns true.
+        bool expiredBad() const
+        {
+            return nExpired > 0;
+        }
+
+        bool reuseBad() const
+        {
+            return nPwdsReused > nPwdsUnique / 10;
+        }
+
+        bool maxReuseBad() const
+        {
+            return maxPwdReuse() > 3;
+        }
+
+        bool avgLenBad() const
+        {
+            return pwdAvgLen() < 10;
+        }
+
+    private:
+        std::map<QString, int> pwds_;
+
+        void countGroup(const Group& group)
+        {
+
+            // Skip the recycle bin
+            if (&group == group.database()->metadata()->recycleBin())
+                return;
+
+            // Handle this group
+            ++nGroups;
+            nEntries += group.entries().size();
+
+            // Handle the entries of this group
+            for (const auto& entry : group.entries()) {
+
+                if (entry->isExpired())
+                    ++nExpired;
+
+                // Get password statistics
+                const auto pwd = entry->password();
+                if (!pwd.isEmpty()) {
+                    if (pwds_.find(pwd) == pwds_.end()) {
+                        ++nPwdsUnique;
+                    } else {
+                        ++nPwdsReused;
+                    }
+
+                    pwdTotalLen += pwd.size();
+
+                    pwds_[pwd]++;
+                }
+            }
+
+            // Recursively handle sub-groups
+            const auto& children = group.children();
+            for (const auto child : children) {
+                if (child) {
+                    countGroup(*child);
+                }
+            }
+        }
+    };
+} // namespace
+
+DatabaseSettingsWidgetStatistics::DatabaseSettingsWidgetStatistics(QWidget* parent)
+    : QWidget(parent)
+    , m_ui(new Ui::DatabaseSettingsWidgetStatistics())
+{
+    m_ui->setupUi(this);
+}
+
+DatabaseSettingsWidgetStatistics::~DatabaseSettingsWidgetStatistics()
+{
+}
+
+void DatabaseSettingsWidgetStatistics::loadSettings(QSharedPointer<Database> db)
+{
+    m_referencesModel.reset(new QStandardItemModel());
+    m_referencesModel->setHorizontalHeaderLabels(QStringList() << tr("Name") << tr("Value"));
+
+    auto errIcon = FilePath::instance()->icon("status", "dialog-error");
+
+    const auto show = [this, &errIcon](QString name, QString value, bool bad = false, QString badMsg = "") {
```

Comment:
This should be named "addStatsRow" and I would make it a private function not a lambda.